### PR TITLE
Hint Generation Protocol

### PR DESCRIPTION
**Pull request by Sumukh**
Changes to Ok:
* The analytics protocol now keeps track of the number of attempts per question and overall. This info is included in the analytics messages (and stored locally as `.ok_history`) 
   * `analytics` should be run after the `grading` protocol for full functionality

Additions:
* Hint generation protocol: It queries the server for changes and displays hints along the lines of: 

> The computer thinks that the conditional expression on line 4 has some errors.

It also provides rubber duck debugging prompts to help guide students to solve questions themselves. 

The prompts are along the lines of:
> What kinds of debugging strategies have you tried? [List of strategies]: 
> Could you describe what the error is for the computer?: 

(after hint is displayed and student corrects error) 
> Could you summarize for the computer what you had to change to fix the bug? 
> If another student had the same error you had, what advice would you give them? 
> How helpful was the hint that the computer provided? [1 as not helpful, 5 as very helpful]: 

For code review: 
- I'd appreciate comments on the changes I've made to the analytics protocol. 
- It's likely that the hint protocol won't be used in this form for long (the plan is that it is used for one assignment), so maintenance of that protocol isn't too big of a concern)

### CODE REVIEW

File: client/protocols/hinting.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+#####################
+# Hinting Mechanism #
+#####################
+
+class HintingProtocol(protocol_models.Protocol):
+    """A protocol that provides rubber duck debugging and hints if applicable.
+    """
+
+    HINT_SERVER = "http://146.148.59.194:5000/"
```
Comment (Sumukh) [inline client/protocols/hinting.py]: Will be HTTPS and use an actual domain soon.

**Comment by Sumukh** [comment]
Sample:
Sample of hint generation after 30+ attempts at a question
```
ok -q nodots
=====================================================================
Assignment: Homework 6
OK, version v1.5.5

Running tests

nodots > Suite 1 > Case 3

scm> (load 'hw06)

scm> (nodots '((1 . 2) 3))
((1 . 2) 3)

# Error: expected
#     ((1 2) 3)
# but got
#     ((1 . 2) 3)

---------------------------------------------------------------------
Test summary
    2 test cases passed before encountering first failed test case

Check for hints on nodots? [yes/no]: yes
-- While we wait, respond to this question. When you are done typing, press Enter.
Could you describe what the function you are working is supposed to do at a high level?

Your Response: Turn malformed lists into well formed lists

The computer thinks that:
1. The conditional expression on line 35 has an error.
```

^ That is an example of heavy guidance because the hint says what to do.
Lighter guidance is along the lines of "Inspect the conditional expressions"

**Comment by Sumukh** [comment]
@albert12132 or @brianhou - Any comments? 

We're going to need this tomorrow so I'm going to merge & deploy before then.

File: client/protocols/analytics.py [hunk 2]
```diff
@@ -89,5 +96,87 @@ def replaced(self, contents):
...
+                     ', assume no history')
+        return history
+
+    def log_run(self, messages):
+        """Record this run of the autograder to a local file.
+        """
+        history = self.read_history()
+        history['all_attempts'] += 1
+        analytics = messages['analytics']
+        questions = analytics.get('question', [])
```
Comment (albert12132) [inline client/protocols/analytics.py]: You should add some comments about each of these variables and what they're supposed to represent/what their data types are. Just looking at the variable names doesn't make it immediately clear what they represent.

File: client/protocols/analytics.py [hunk 3]
```diff
@@ -89,5 +96,87 @@ def replaced(self, contents):
...
+            pickle.dump(history, f)
+            os.fsync(f)
+
+        messages['analytics']['history'] = history
+
+def is_correct(score):
+    """Given a score from the grading protocol, see if no failed cases
+    and no locked cases.
+    """
+    return sum(score.values()) == score['passed']
```
Comment (albert12132) [inline client/protocols/analytics.py]: I'd have thought a score was just a single number, so I'm surprised to see that it's a dictionary. Please add a description in the docstring of what score's datatype is and what info it contains.

Comment (albert12132) [inline client/protocols/analytics.py]: And perhaps "score" isn't the best name for this...perhaps `grading_results`?

File: client/protocols/analytics.py [hunk 4]
```diff
@@ -89,5 +96,87 @@ def replaced(self, contents):
...
+
+def first_failed_test(tests, scores):
+    names = [t.name for t in tests]
+    for test in names:
```
Comment (albert12132) [inline client/protocols/analytics.py]: I'd prefer replacing `test` with `test_name` or `name`.

File: client/protocols/analytics.py [hunk 5]
```diff
@@ -89,5 +96,87 @@ def replaced(self, contents):
...
+    def log_run(self, messages):
+        """Record this run of the autograder to a local file.
+        """
+        history = self.read_history()
+        history['all_attempts'] += 1
+        analytics = messages['analytics']
+        questions = analytics.get('question', [])
+        grading = 'grading' in messages and messages['grading']
+
+        # Attempt to figure out what question is being worked on
```
Comment (albert12132) [inline client/protocols/analytics.py]: Please elaborate more on what "worked on" means, e.g. "the student is currently solving/implementing"

Comment (albert12132) [inline client/protocols/analytics.py]: And add a note that the assumption is, if a student is currently working on question N, they will have finished questions 1 through N - 1, so the tests for question N will be the first ones to fail.

File: client/protocols/analytics.py [hunk 6]
```diff
@@ -89,5 +96,87 @@ def replaced(self, contents):
...
+            # Update earlier question correctness status
+            for saved_q, details in history['questions'].items():
+                finished = details['solved']
+                if not finished and saved_q in grading:
+                    score = grading[saved_q]
+                    details['solved'] = is_correct(score)
+        else:
+            history['question'] = questions
+
+        for question in questions:
```
Comment (albert12132) [inline client/protocols/analytics.py]: Add a comment explaining what this for loop does on a high level, since this for loop is rather involved.

File: client/protocols/hinting.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+        current_q = history['question']
+
+        for question in current_q:
+            if question not in questions:
+                continue
+            stats = questions[question]
+            messages['hinting'][question] = {'prompts': {}, 'reflection': {}}
+            hint_info = messages['hinting'][question]
+
+            # Only prompt
```
Comment (albert12132) [inline client/protocols/hinting.py]: Not sure what this means. Please elaborate.

File: client/protocols/hinting.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+            if (stats['solved'] and stats['attempts'] > 6):
```
Comment (albert12132) [inline client/protocols/hinting.py]: Parentheses not needed.

Comment (albert12132) [inline client/protocols/hinting.py]: The "6" should not be hardcoded here.

File: client/protocols/hinting.py [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+                if self.args.question:
+                    # Only prompt for reflection with question specified.
+                    log.info('Giving reflection response on %s', question)
+                    reflection = random.choice(SOLVE_SUCCESS_MSG)
+                    if not confirm("Nice work! Could you answer a quick question"
+                                   " about how you approached this question?"):
+                        hint_info['reflection']['accept'] = False
+                    else:
+                        hint_info['reflection']['accept'] = True
+                    prompt_user(reflection, hint_info)
```
Comment (albert12132) [inline client/protocols/hinting.py]: Just to clarify, are you prompting the user even if they refuse to answer a quick question?

File: client/protocols/hinting.py [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+            elif (stats['attempts'] < 5):
```
Comment (albert12132) [inline client/protocols/hinting.py]: Don't hardcode this 5 here.

File: client/protocols/hinting.py [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+            else:
+                hint_info['elgible'] = not stats['solved']
+
+            if not hint_info['elgible']:
+                continue
+            log.info('Prompting for hint on %s', question)
+
+            if confirm("Check for hints on {}?".format(question)):
+                hint_info['accept'] = True
+                print("Thinking... (could take upto 15 seconds)", end='')
```
Comment (albert12132) [inline client/protocols/hinting.py]: "upto" -> "up to"

File: client/protocols/hinting.py [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+                    hint = response['message']
+                    pre_prompt = response['pre-prompt']
+                    post_prompt = response['post-prompt']
+                    log.info("Hint server response: {}".format(response))
+                    if not hint and not pre_prompt:
+                        print("\nSorry. No hints found for {}".format(question))
+                        continue
+
+                    if pre_prompt:
+                        print("\r-- While we wait, respond to this question."
```
Comment (albert12132) [inline client/protocols/hinting.py]: What exactly are we waiting for? Won't we only get to this code path when `query_Server` finishes executing?

File: client/protocols/hinting.py [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+                    print("\r\nCould not get a hint.")
+            else:
+                log.info('Declined Hints for %s', question)
+                hint_info['accept'] = False
+
+    def query_server(self, messages, test):
+        access_token, _, _ = auth.get_storage()
+        user = auth.get_student_email(access_token) or access_token
+        if user:
+            user = hash(user)
```
Comment (albert12132) [inline client/protocols/hinting.py]: Please add a comment explaining why this is hashed.

File: client/protocols/hinting.py (hunk 12 above)
Comment (Sumukh) [inline client/protocols/hinting.py]: You are correct. It was supposed to be intentionally deceptive - but it's not necessary to be deceptive here. I'll update the message.

File: client/protocols/hinting.py [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+        request.add_header("Content-Type", "application/json")
+
+        response = urllib.request.urlopen(request, serialized_data, 10)
+        return json.loads(response.read().decode('utf-8'))
+
+def prompt_user(query, results):
+    try:
+        response = None
+        short_resps = 0
+        while not response or response == '':
```
Comment (albert12132) [inline client/protocols/hinting.py]: response == '' is redundant.

File: client/protocols/hinting.py [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+            response = input("{}\nYour Response: ".format(query))
+            if not response or response == '':
+                short_resps += 1
```
Comment (albert12132) [inline client/protocols/hinting.py]: Why is this called "short"?

File: client/protocols/hinting.py [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+                if short_resps > 2:
```
Comment (albert12132) [inline client/protocols/hinting.py]: Explain why you are breaking here (e.g. "don't ask more than 2 times, to avoid annoying the user")

File: client/protocols/hinting.py [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+            print("Exiting Hint") # Second I/O will get KeyboardInterrupt
+            return ''
+        except KeyboardInterrupt:
+            return ''
+
+def confirm(message):
+    response = input("{} [yes/no]: ".format(message))
+    return response.lower() == "yes" or response.lower() == "y"
+
+
```
Comment (albert12132) [inline client/protocols/hinting.py]: There should only be up to 2 blank lines following this function.

File: client/protocols/hinting.py (hunk 11 above)
Comment (albert12132) [inline client/protocols/hinting.py]: And why is the end character changed to '', as opposed to using a newline?

Comment (Sumukh) [inline client/protocols/hinting.py]: The idea is that the next print uses `\r` to clear the line. If we terminate the line we can't clear this line. 

In my experience it's been a little iffy at actually displaying this line if it doesn't end in a new line - so I'll change it back to end in a new line and avoid the line clearing business.

**Comment by Sumukh** [comment]
@albert12132 / @brianhou  - Any other comments? I want to deploy this soon so that I can launch a 61A Homework.

**Comment by albert12132** [comment]
LGTM.
